File: xjakt00-adventura/PrikazKup.java
/*******************************************************************************
 * Příkaz 'kup' umožňuje hráči nakupovat věci v obchodech.
 *
 * @author    Tomáš Jakubčin (xjakt00)
 * @version   pro školní rok 2010/11
 */
public class PrikazKup implements IPrikaz
{
    private static final String NAZEV = "kup";
    private Batoh batoh;
    private HerniPlan plan;
    private Hra hra;

    /***************************************************************************
     *  Konstruktor třídy.
     *  
     *  @param  plan    herni plan.
     *  @param  batoh   odkaz na hráčův batoh.
     *  @param  hra     odkaz na hru
     */
    public PrikazKup(Batoh batoh, HerniPlan plan, Hra hra) {
        this.batoh = batoh;
        this.plan = plan;
        this.hra = hra;
    }

    /***************************************************************************
     *  Nejdříve kontroluje, jestli je zadný parametr a pak všechny další náležitosti -
     *  např. jestli se vejde do batohu apod. V případě, že všechno dobře dopadne, přidá
     *  hráči koupený předmět do batohu a odebere mu peníze. Nedá se uplatnit na věci, které
     *  jsou položeny.
     */
    public String proved (String... parametry) {
        if (parametry.length == 0) {
            return "Nevim, co mam koupit.";
        }
        Prostor kdeJsme = plan.getAktualniProstor();
        if (kdeJsme.getObchod() == false) {
            return "Nemuzes koupit vec, ktera neni v obchode. Pouzij prikaz vezmi.";
        }
        String nazevCoKoupit = parametry [0];
        Vec vecicka = kdeJsme.vyberVec(nazevCoKoupit);
        if (vecicka == null)
            return "Tato vec v prostoru neni.";
        if (vecicka.isPrenosna() == false)
            return "Tuto vec nemuzes koupit, neni prenosna.";
        if (batoh.getPocetVeci()+1 > batoh.MAX_VECI)
            return "Tato vec se jiz do batohu nevejde.";
        if (batoh.getVaha()+vecicka.getVaha() > batoh.MAX_VAHA)
            return "Tato vec je moc tezka, vyprazdni nejdrive batoh.";
        if (vecicka.getCena() > batoh.getPenize())
            return "Na tuto vec nemas dostatek penez.";
        if (plan.getCas()-1 > 0) {
            plan.setCas(1);
            vecicka.setVBatohu(1);
            batoh.setPocetVeci(1);
            batoh.setVaha(vecicka.getVaha());
            batoh.setPenize(vecicka.getCena());
            batoh.setVec(vecicka);
            return "Vec '"+nazevCoKoupit+"' byla koupena a nachazi se nyni ve tvem batohu.";
        }
        else {
            plan.konecHry();
            return "";
        }
    }

    /***************************************************************************
     *
     */
    public String getNazev() {
        return NAZEV;
    }
}

File: src/Adventura/PrikazKup.cpp
#include "PrikazKup.h"

#include "Hra.h"

PrikazKup::PrikazKup(const std::string & nazev, Hra * hra, HerniPlan * herni_plan, Batoh * batoh)
	: IPrikaz(nazev), odkaz_na_hru(hra), odkaz_na_plan(herni_plan), odkaz_na_batoh(batoh)
{
}

PrikazKup::~PrikazKup()
{
}


void PrikazKup::proved(const std::vector<std::string> & parametry)
{
	if(parametry.empty())
	{
		std::cout << "Nevim co mam koupit.\n";
	}
	else if (parametry.size() == 1)
	{
		auto aktualni_prostor = odkaz_na_plan->get_aktualni_prostor();
		if (aktualni_prostor->je_obchod() == true)
		{
			if (aktualni_prostor->je_vec_v_prostoru(parametry[0]) == false)
			{
				std::cout << "Tato vec v obchodu neni, pokud je na zemi neco polozeno, pouzij prikaz 'vezmi'.\n";
				return;
			}

			int cena = aktualni_prostor->get_cena(parametry[0]);
			if (cena > odkaz_na_batoh->get_penize())
			{
				std::cout << "Na tuto vec nemas dostatek penez.\n";
				return;
			}
			else
			{
				aktualni_prostor->prodej_vec(parametry[0]);
				odkaz_na_batoh->pridej_vec(parametry[0]);
				odkaz_na_batoh->pridej_penize(-cena);

				std::cout << "Vec " << parametry[0] << " byla koupena a nachazi se nyni ve tvem batohu.\n";
				return;
			}
		}
		else
		{
			std::cout << "Tento prostor neni obchod, pokud je na zemi neco polozeno, pouzij prikaz 'vezmi'.\n";
			return;
		}
	}
	else
	{
		std::cout << "Neni mozne koupit dve veci zaroven.\n";
	}
}

std::string PrikazKup::nazev(void) const
{
	return IPrikaz::jmeno_prikazu;
}

std::string PrikazKup::pouziti(void)const
{
	return "kup <co>";
}